Update 06-Test_HTTP_Methods (#531)
OWASP/wstg: document/4-Web_Application_Security_Testing/02-Configuration_and_Deployment_Management_Testing/06-Test_HTTP_Methods.md

When testing an application that has to accept other methods, e.g. a RESTful Web Service, test it thoroughly to make sure that all endpoints accept only the methods that they require.

> Audit Guideline for HTTP PUT method enabled

Reviewer: This should be a level 4 heading and following similar naming as other sections.

Author: Done

Reviewer: No it's not?

> 1) Capture the base request of the vulnerable website with a personal proxy and leverage that request via the proxy's resend functionality.

Reviewer: 1. Please use markdown numbering (ordered list).
2. How do we know the website is vulnerable at this point?

Author: Done

Reviewer: No it's not?

> 2) Change the request method to PUT and set the path along the shell as follow and send the request to the application server.

Reviewer: PUT should use code fencing (backticks).
I'm not sure what "set the path along the shell as follow" means.

> `/test/shell.php` for PHP based server or `/test/shell.jsp ` for JAVA based server or `/test/shell.aspx` for ASP.NET based server 

Reviewer: It would probably be best to have an actual example request here so that it's clear the to tester what's a local path and what's a remote path, etc.

Also it might be worth trying something innocuous before trying a web shell, being able to upload a text file, image, etc can still be an issue in the end while more "active" content might be blocked.

> 3) Observe the response if the server response with 201 Created response. Then the application is vulnerable.

Reviewer: I think this might make a poor assumption.
What if it's 200?
What if it's a redirect to the object that was just uploaded?

> 4) Run the command on remote application server.

Reviewer: This assumes that not only did you manage to PUT a web shell but that it was treated appropriately without any further efforts on the part of the tester. Which may or may not be true. PUTting a web shell isn't the only issue related to successful PUT requests.

> Note- If the HTTP PUT method is not allowed on base URL or request, try uploading on a different directory in the application.

Reviewer: ```suggestion
> Note If the HTTP `PUT` method is not allowed on base URL or request, try other paths in the system.
```

When testing an application that has to accept other methods, e.g. a RESTful Web Service, test it thoroughly to make sure that all endpoints accept only the methods that they require.

> #### Audit Guideline for HTTP `PUT` method enabled

Reviewer: ```suggestion
#### Testing the PUT Method
```

I know I said something earlier about backticks and PUT, but it is problematic for headings which are used elsewhere as fragment anchors.

1. Capture the base request of the testing website with a personal proxy and leverage that request via the proxy's resend functionality.
2. Change the request method to `PUT` and add `test.html` file and send the request to the application server.

   ```html
   PUT /test.html HTTP/1.1
   Host: testing-website

   <html>
   HTTP PUT Method is Enabled
   </html>
   ```

3. If the server response with 2XX success codes or 3XX redirections and then confirm by `GET` request for `test.html` file. The application is vulnerable.

If the HTTP `PUT` method is not allowed on base URL or request, try other paths in the system.

> 

Reviewer: It is still worth mentioning web shells here. A verbose description is probably more appkicable to the file upload or RFI section(s) but a line or two here to keep testers thinkung woukd be a good move I think.

Author: check. is it ok?

When testing an application that has to accept other methods, e.g. a RESTful Web Service, test it thoroughly to make sure that all endpoints accept only the methods that they require.

> #### Testing the PUT Method

Reviewer 2: Isn't this better as "Testing File Upload"? POST and PUT can both do this, just needs a different body.
If PUT is allowed to upload files, then POST most probably allows to override them.
I would recommend setting 2 subsections, one for POST and one for PUT.

Author: Hi ThunderSon,

I don't think, we should see this as "Testing File Upload". We can definitely upload a file, for sure. But, now we should look it with the perspective if the application owner or developer left HTTP PUT method enabled.

Reviewer 2: It's a similar case, which can happen in both cases if POST and PUT are not properly configured. I care about having a nice flow to what's actually happening. I'll leave @kingthorin to give the judgment on this :)

Reviewer: This is supposed to be specific to method testing. There is a separate section that covers file upload testing.

> 1. Capture the base request of the testing website with a personal proxy and leverage that request via the proxy's resend functionality.

Reviewer 2: ```suggestion
1. Capture the base request of the target with a web proxy.
```

2. Change the request method to `PUT` and add `test.html` file and send the request to the application server.

   ```html
   PUT /test.html HTTP/1.1
   Host: testing-website

   <html>
   HTTP PUT Method is Enabled
   </html>
   ```

> 3. If the server response with 2XX success codes or 3XX redirections and then confirm by `GET` request for `test.html` file. The application is vulnerable. 

Reviewer 2: ```suggestion
3. If the server responds with a success code (2XX) or a redirection code (3XX), confirm the creation of the file by issuing a `GET` request for `test.html` where the server is storing it. Sometimes you'll have to identify the location where it is being stored which can be done by dirbusting the web application. If the server returns that file, the application is vulnerable to file upload and code injection.
```

4. Now, add the webshell using the `PUT` method.
   
   `shell.php` for PHP based server or `shell.jsp` for JAVA based server or `shell.aspx` for ASP.NET based server

> If the HTTP `PUT` method is not allowed on base URL or request, try other paths in the system.

Reviewer 2: In my opinion, this shouldn't be escalated to a shell unless the attacker is after a deeper assessment. Grabbing a shell for "proof" isn't realistic. A tester shouldn't make a web application vulnerable in any way. Just uploading a file should be enough to prove that this is vulnerable.

Reviewer: From my POV there are a number of potential scenarios that arise related to HTTP PUT.
1. Can the tester PUT something and exhaust/waste resources?
2. Can the tester PUT something replacing existing content?
3. Can the tester PUT something that is then served as active web content (PHP, JSP, ASP, etc) that can then be abused to further harm/manipulate/steal resources/data.

Having shells mentioned as a 4th step probably isn't necessary but having something that mentions the potential for something more serious than resource exhaustion/DoS seems valid.

I'm being long winded....sorry. Basically drop step 4 (Lines 60-21) and stick with line 68 😉

Reviewer 2: Nice to mention as well that sometimes you could upload benign text, sometimes have execution, and sometimes set system files, which can be merged with your points.

> NOTE: If you are successful in uploading a web shell you should overwrite it or ensure that the security team of the target are aware and remove the component promptly after your proof-of-concept.

> An attacker may be able to place arbitrary, and potentially malicious, content into the application which may lead to RCE, defacing the application.

Reviewer: ```suggestion
Leveraging the `PUT` method an attacker may be able to place arbitrary and potentially malicious content, into the system which may lead to remote code execution, or defacing the site.
```
